### ISSUE TITLE
decoding valid UTF-16 (to UTF-8) panics with an output buffer of size 4

### ISSUE THREAD

**Issue by BurntSushi** [hsivonen/encoding_rs#34]
Here's the full program:

```rust
extern crate encoding_rs;

fn main() {
    let enc = encoding_rs::UTF_16LE;
    let mut decoder = enc.new_decoder_with_bom_removal();

    let mut dst = [0u8; 4];

    let src = &[9];
    let (res, nin, nout, err) = decoder.decode_to_utf8(src, &mut dst, false);
    eprintln!(
        "res: {:?}, nin: {:?}, nout: {:?}, err: {:?}",
        res, nin, nout, err,
    );
    assert_eq!(dst, [0x00, 0x00, 0x00, 0x00]);

    let src = &[60, 103, 62];
    let (res, nin, nout, err) = decoder.decode_to_utf8(src, &mut dst, false);
    eprintln!(
        "res: {:?}, nin: {:?}, nout: {:?}, err: {:?}",
        res, nin, nout, err,
    );
    assert_eq!(dst, [0xE6, 0xB0, 0x89, 0x00]);
}
```

And the output:

```
[andrew@Cheetah encodingrsbug]$ cargo run
    Finished dev [unoptimized + debuginfo] target(s) in 0.01s
     Running `target/debug/encodingrsbug`
res: InputEmpty, nin: 1, nout: 0, err: false
thread 'main' panicked at 'attempt to subtract with overflow', /home/andrew/.cargo/registry/src/github.com-1ecc6299db9ec823/encoding_rs-0.8.6/src/handles.rs:277:31
note: Run with `RUST_BACKTRACE=1` for a backtrace.

[andrew@Cheetah encodingrsbug]$ cargo run --release
   Compiling encodingrsbug v0.1.0 (/home/andrew/tmp/ripgrep/issues/1052/encodingrsbug)
    Finished release [optimized] target(s) in 0.38s
     Running `target/release/encodingrsbug`
res: InputEmpty, nin: 1, nout: 0, err: false
thread 'main' panicked at 'index out of bounds: the len is 1 but the index is 1', /home/andrew/.cargo/registry/src/github.com-1ecc6299db9ec823/encoding_rs-0.8.6/src/handles.rs:309:21
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

This _looks_ like the same symptom as #32, however, I filed a new issue since the input data is different, and because I actually seem to need two `decode_to_utf8` calls here to trigger the panic. Namely, if I prepend `9` to the second `src` buffer and remove the first call to `decode_to_utf8`, then the program works as expected. Moreover, I _believe_ the program in #32 is actually invalid since the docs for this crate state that the output buffer for UTF-8 must be at least 4 bytes, and the buffer size in #32 is 2 bytes, so it seems like a panic there is legal. However, in this program, the output buffer size is 4 bytes, which I _think_ makes this a correct program with respect to satisfying the preconditions of `decode_to_utf8`. However, I am not 100% sure.

If this is a correct program, then I think the panic is a bug. I briefly looked at the code but couldn't immediately see the appropriate fix.

If this is an incorrect program, then I think that provokes a question: is it possible to write a correct program that uses `encoding_rs` with a statically known fixed size buffer? If so, what is the minimum size of said buffer?

To give you context, I tripped this bug in `encoding_rs_io`. Specifically, this is code that is handling the implementation of `Read` when the caller provided buffer is too small to fit a single decoded codepoint. Namely, it falls over to to the "tiny" transcoder: https://github.com/BurntSushi/encoding_rs_io/blob/a322c14c9ea48303fb883668a605f99aaf734357/src/util.rs#L6-L22

cc @sinkuu

**Comment by hsivonen** [comment]
> Moreover, I believe the program in #32 is actually invalid since the docs for this crate state that the output buffer for UTF-8 must be at least 4 bytes, and the buffer size in #32 is 2 bytes, so it seems like a panic there is legal.

The implementation is not supposed to panic in that case but to return a status that fails to make progress. (I.e. it should be OK to pass a short tail slice as part of a loop that started with a longer buffer, but the loop will be infinite if the intermediate buffer as a whole isn't at least 4 bytes such that some iterations offer at least 4 bytes to write to.)

> If this is a correct program, then I think the panic is a bug.

Your program looks correct. I'll investigate. (Might take until next week, though. Apologies in advance.)

**Comment by BurntSushi** [comment]
No worries! Thanks for looking into it! I worked around this for now by just increasing the buffer size by a few bytes. (Of course, I don't know if that just makes the bug harder to trip though.)

**Comment by hsivonen** [comment]
Fixed in 0.8.7. Thanks for catching this!

**Comment by hsivonen** [comment]
[Gecko bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1494651)
